pkg/cache/client/context.go

```diff
@@ -0,0 +1,45 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+
+	"github.com/kcp-dev/kcp/pkg/cache/client/shard"
+)
+
+type shardKey int
+
+const (
+	// shardContextKey is the context key for a request
+	shardContextKey shardKey = iota
+)
+
+// WithShardInContext returns a context with the given shard set
+func WithShardInContext(parent context.Context, shard shard.Name) context.Context {
+	return context.WithValue(parent, shardContextKey, shard)
+}
+
+// ShardFromContext returns the value of the shard key on the ctx,
+// or an empty Name if there is no shard key.
+func ShardFromContext(ctx context.Context) shard.Name {
+	val, ok := ctx.Value(shardContextKey).(shard.Name)
```

Comment:
nit: can be a follow-up: can just `return ctx.Value(shardContextKey).(shard.Name)` since 

a) the `shardContextKey` is private to this package, so the only case where `!ok` is if someone were to write code _in this package_ to add something invalid
b) the zero-value is already `shard.New("")`

Reply:
yes, I just don't like when the code panics but sure I can change it.

Reply:
done.